src/main/kotlin/com/terraformation/backend/accelerator/db/ProjectScoreStore.kt

```diff
@@ -0,0 +1,100 @@
+import com.terraformation.backend.customer.model.requirePermissions
+import com.terraformation.backend.db.accelerator.CohortPhase
+import com.terraformation.backend.db.accelerator.tables.references.COHORTS
+import com.terraformation.backend.db.accelerator.tables.references.PARTICIPANTS
+import com.terraformation.backend.db.accelerator.tables.references.PROJECT_SCORES
+import com.terraformation.backend.db.asNonNullable
+import com.terraformation.backend.db.default_schema.ProjectId
+import com.terraformation.backend.db.default_schema.tables.references.PROJECTS
+import jakarta.inject.Named
+import java.time.InstantSource
+import org.jooq.DSLContext
+
+@Named
+class ProjectScoreStore(
+    private val clock: InstantSource,
+    private val dslContext: DSLContext,
+) {
+  /**
+   * Returns the per-category scores for a project.
+   *
+   * @param phases If non-null, only return scores for these phases.
+   */
+  fun fetchScores(
+      projectId: ProjectId,
+      phases: Collection<CohortPhase>? = null
+  ): Map<CohortPhase, List<ExistingProjectScoreModel>> {
+    requirePermissions { readProjectScores(projectId) }
+
+    return with(PROJECT_SCORES) {
+      val conditions =
+          listOfNotNull(
+              phases?.let { PHASE_ID.`in`(it) },
+              PROJECT_ID.eq(projectId),
+          )
+
+      dslContext
+          .select(PHASE_ID, MODIFIED_TIME, QUALITATIVE, SCORE, SCORE_CATEGORY_ID)
+          .from(PROJECT_SCORES)
+          .where(conditions)
+          .orderBy(PHASE_ID, SCORE_CATEGORY_ID)
+          .fetchGroups(PHASE_ID.asNonNullable()) { ProjectScoreModel.of(it) }
+    }
+  }
+
+  fun updateScores(
+      projectId: ProjectId,
+      phase: CohortPhase,
+      scores: Collection<NewProjectScoreModel>
+  ) {
+    requirePermissions { updateProjectScores(projectId) }
+
+    val now = clock.instant()
+    val userId = currentUser().userId
+
+    val currentPhase =
+        dslContext
+            .select(COHORTS.PHASE_ID)
+            .from(PROJECTS)
+            .join(PARTICIPANTS)
+            .on(PROJECTS.PARTICIPANT_ID.eq(PARTICIPANTS.ID))
+            .join(COHORTS)
+            .on(PARTICIPANTS.COHORT_ID.eq(COHORTS.ID))
+            .fetchOne(COHORTS.PHASE_ID) ?: throw ProjectNotInCohortException(projectId)
+
+    if (currentPhase != phase) {
+      throw ProjectNotInCohortPhaseException(projectId, phase)
+    }
+
+    dslContext.transaction { _ ->
+      scores.forEach { score ->
+        with(PROJECT_SCORES) {
+          dslContext
+              .insertInto(PROJECT_SCORES)
+              .set(CREATED_BY, userId)
+              .set(CREATED_TIME, now)
+              .set(MODIFIED_BY, userId)
+              .set(MODIFIED_TIME, now)
+              .set(PHASE_ID, phase)
+              .set(PROJECT_ID, projectId)
+              .set(QUALITATIVE, score.qualitative)
+              .set(SCORE, score.score)
+              .set(SCORE_CATEGORY_ID, score.category)
+              .onConflict()
```

Comment:
Does this work the same as `onDuplicateKeyUpdate`?

Reply:
The "on conflict" syntax is the native PostgreSQL way to do this. "On duplicate key" is the MySQL syntax (maybe MySQL picked it up from somewhere else originally, not sure) but jOOQ translates it to an equivalent PostgreSQL statement at runtime.